packages/package-manager/src/YarnPackageManager.ts

```diff
@@ -0,0 +1,139 @@
+import spawnAsync, { SpawnOptions } from '@expo/spawn-async';
+import ansiRegex from 'ansi-regex';
+import assert from 'assert';
+import fs from 'fs';
+import path from 'path';
+import rimraf from 'rimraf';
+import { Transform } from 'stream';
+
+import { DISABLE_ADS_ENV } from './NodePackageManagers';
+import { Logger, PackageManager } from './PackageManager';
+import isYarnOfflineAsync from './utils/isYarnOfflineAsync';
+
+const ansi = `(?:${ansiRegex().source})*`;
+const yarnPeerDependencyWarningPattern = new RegExp(
+  `${ansi}warning${ansi} "[^"]+" has (?:unmet|incorrect) peer dependency "[^"]+"\\.\n`,
+  'g'
+);
+
+/** Exposed for testing */
+export class YarnStderrTransform extends Transform {
+  _transform(
+    chunk: Buffer,
+    encoding: string,
+    callback: (error?: Error | null, data?: any) => void
+  ) {
+    this.push(chunk.toString().replace(yarnPeerDependencyWarningPattern, ''));
+    callback();
+  }
+}
+
+export class YarnPackageManager implements PackageManager {
+  options: SpawnOptions;
+  private log: Logger;
+
+  constructor({ cwd, log, silent }: { cwd: string; log?: Logger; silent?: boolean }) {
+    this.log = log || console.log;
+    this.options = {
+      env: {
+        ...process.env,
+        ...DISABLE_ADS_ENV,
+      },
+      cwd,
```

Comment:
it's not clear to me what cwd is, if I saw an interface like this (and I know that this package supports workspaces), I would expect that if cwd is in a subproject most command will be executed there, but e.g. yarn install would run in root.

Reply:
I think, but correct me if I'm wrong, I addressed this in https://github.com/expo/expo/pull/18576/files#diff-674041de5e2d6dc1344c60e176ff815c71ad2c76c1ac823e2c99b77022813c4fR5-R19. There is only 1 `PackageManagerOptions`, which inherits the `SpawnOptions` (with additional `silent` and `log` options). I think that should make it a bit more clear.

Besides that, there is also the `packageManager.workspaceRoot()` which returns a package manager instance of the same type (`npm|pnpm|yarn`), only if there was a workspace root detected. You can use that to run install from root:

```
// Compact version
await (packageManager.workspaceRoot() ?? packageManager).installAsync()

// Example with logging
const workspaceManager = packageManager.workspaceRoot();
if (workspaceManager) {
  console.log(`Running install in workspace root ${workspaceManager.options.cwd}`);
  await workspaceManager.installAsync();
} else {
  console.log(`Running install in project ${packageManager.options.cwd}`);
  await packageManager.installAsync();
}
```